### PR TITLE
feat: Teach InfluxQL how to plan `LIMIT` and `OFFSET` with `GROUP BY`

### LINKED ISSUE #6920: InfluxQL-IOx: Teach planner how to process `LIMIT` and `OFFSET` clauses
Teach the InfluxQL query planner how to generate a logical plan for the `LIMIT` and `OFFSET` clauses.

# Background

The `LIMIT` clause restricts the number of rows returned _for each group_. The `OFFSET` clause specifies how many
rows to skip, _for each group_, before returning results.

Familiarisation of the **Data Flow - Grouping Behaviour** section of #6112  is essential in understanding the group key concept and how points are grouped in InfluxQL.

The group for `LIMIT` and `OFFSET` is made up of the

* measurement name, and
* tags from the `GROUP BY`

If there is no `GROUP BY tag`, then the group is equivalent to just the measurement. This means the points cannot be aggregated or combined across measurements.

## Examples

The `LIMIT` and `OFFSET` behaviour is illustrated with the following queries.

The simplest example is a raw query[^raw], with no `GROUP BY` clause and querying a single measurement:

```sql
select cpu, usage_idle FROM cpu LIMIT 3;
```
```
name: cpu
time                 cpu       usage_idle
----                 ---       ----------
2020-06-11T16:52:00Z cpu-total 97.93916821849783
2020-06-11T16:52:00Z cpu0      89.56262425447316
2020-06-11T16:52:00Z cpu1      99.8015873015873
```

The equivalent DataFusion SQL for this simple case is quite similar[^1]:

```sql
select 'cpu' AS measurement, time, cpu, usage_idle from cpu order by time, cpu limit 3;
```
```
+-------------+----------------------+------+-------------------+
| measurement | time                 | cpu  | usage_idle        |
+-------------+----------------------+------+-------------------+
| cpu         | 2020-06-11T16:52:00Z | cpu4 | 94.93545183714002 |
| cpu         | 2020-06-11T16:52:00Z | cpu1 | 99.8015873015873  |
| cpu         | 2020-06-11T16:52:00Z | cpu5 | 99.80139026812314 |
+-------------+----------------------+------+-------------------+
```

Adding a `GROUP BY`, and a `WHERE` clause to keep the result set small, demonstrates how `LIMIT` is applied to each group:

```sql
select cpu, usage_idle FROM cpu WHERE cpu =~ /cpu(0|1)$/ GROUP BY cpu LIMIT 3;
```
```
name: cpu
tags: cpu=cpu0
time                 cpu  usage_idle
----                 ---  ----------
2020-06-11T16:52:00Z cpu0 89.56262425447316
2020-06-11T16:52:10Z cpu0 88.6113886113886
2020-06-11T16:52:20Z cpu0 88.6

name: cpu
tags: cpu=cpu1
time                 cpu  usage_idle
----                 ---  ----------
2020-06-11T16:52:00Z cpu1 99.8015873015873
2020-06-11T16:52:10Z cpu1 99.7002997002997
2020-06-11T16:52:20Z cpu1 99.89989989989989
```

The equivalent DataFusion SQL must use window functions to implement the per group limits:

```sql
WITH t AS (select 'cpu' AS measurement, cpu, time, usage_idle, row_number() OVER (PARTITION BY cpu ORDER BY cpu, time) AS row FROM cpu WHERE cpu::varchar ~ 'cpu(0|1)$' ORDER BY cpu, time) SELECT measurement, cpu, time, usage_idle FROM t WHERE CAST(row AS bigint) <= 3;
```

```
+-------------+------+----------------------+-------------------+
| measurement | cpu  | time                 | usage_idle        |
+-------------+------+----------------------+-------------------+
| cpu         | cpu0 | 2020-06-11T16:52:00Z | 89.56262425447316 |
| cpu         | cpu0 | 2020-06-11T16:52:10Z | 88.6              |
| cpu         | cpu0 | 2020-06-11T16:52:20Z | 88.6              |
| cpu         | cpu1 | 2020-06-11T16:52:00Z | 99.8015873015873  |
| cpu         | cpu1 | 2020-06-11T16:52:10Z | 99.8              |
| cpu         | cpu1 | 2020-06-11T16:52:20Z | 99.89989989989989 |
+-------------+------+----------------------+-------------------+
```

Finally, an example of multiple measurements:

```sql
select usage_idle, free from cpu, disk WHERE (cpu =~ /^cpu(0|1)$/ OR fstype = 'apfs') group by cpu, device LIMIT 1
```
```
name: cpu
tags: cpu=cpu0, device=
time                 usage_idle        free
----                 ----------        ----
2020-06-11T16:52:00Z 89.56262425447316

name: cpu
tags: cpu=cpu1, device=
time                 usage_idle       free
----                 ----------       ----
2020-06-11T16:52:00Z 99.8015873015873

name: disk
tags: cpu=, device=disk1s1
time                 usage_idle free
----                 ---------- ----
2020-06-11T16:51:50Z            882941362176

name: disk
tags: cpu=, device=disk1s2
time                 usage_idle free
----                 ---------- ----
2020-06-11T16:51:50Z            882941362176

name: disk
tags: cpu=, device=disk1s5
time                 usage_idle free
----                 ---------- ----
2020-06-11T16:51:50Z            882941362176
```

and the equivalent DataFusion SQL, which is notably far more complex than the InfluxQL version:

```sql
WITH t AS (select 'cpu' as measurement, cpu, NULL as device, time, usage_idle, NULL as free, row_number() OVER (PARTITION BY cpu ORDER BY cpu, time) as row FROM cpu WHERE cpu::varchar ~ '^cpu(0|1)$' UNION ALL select 'disk' as measurement, NULL AS cpu, device, time, NULL as usage_idle, free, row_number() OVER (PARTITION by device ORDER BY device, time) as row FROM disk WHERE fstype::varchar = 'apfs') SELECT measurement, cpu, device, time, usage_idle, free FROM t WHERE row::bigint <= 1 ORDER BY 1, 2, 3, 4;
```
```
+-------------+------+---------+----------------------+-------------------+--------------+
| measurement | cpu  | device  | time                 | usage_idle        | free         |
+-------------+------+---------+----------------------+-------------------+--------------+
| cpu         | cpu0 |         | 2020-06-11T16:52:00Z | 89.56262425447316 |              |
| cpu         | cpu1 |         | 2020-06-11T16:52:00Z | 99.8015873015873  |              |
| disk        |      | disk1s1 | 2020-06-11T16:51:50Z |                   | 882941362176 |
| disk        |      | disk1s2 | 2020-06-11T16:51:50Z |                   | 882941362176 |
| disk        |      | disk1s5 | 2020-06-11T16:51:50Z |                   | 882941362176 |
+-------------+------+---------+----------------------+-------------------+--------------+
```

[^1]: The `measurement` column is added as it is a required part of the group key. It's purpose becomes apparent when querying multiple measurements.
[^raw]: A raw query is one with no aggregates or selectors

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #6920

This PR teaches IOx how to plan InfluxQL queries that include a `LIMIT` and / or `OFFSET` clause when combined with a `GROUP BY` clause. This includes grouping by `TIME` and gap filling along with grouping by tags.

See #6920 for a discussion on the semantics of `LIMIT` and `OFFSET`, which work differently to ANSI SQL.

### CODE REVIEW

**Review by mhilton — APPROVED** [review]

File: iox_query_influxql/src/plan/planner.rs [hunk 1]
```diff
@@ -422,12 +439,110 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
-        )
+        let plan = project(plan, proj_exprs.clone())?;
+
+        let plan = plan_with_sort(plan, select, &group_by_tag_set, &projection_tag_set)?;
+
+        if select.offset.is_some() || select.limit.is_some() {
+            if group_by_tag_set.is_empty() {
+                // If the query is not grouping by tags, the
+                // DataFusion Limit operator is sufficient.
+                self.limit(plan, select.offset, select.limit)
```
Comment (crepererum) [inline iox_query_influxql/src/plan/planner.rs:457]: Does this fancy new sorting only affect select-group-bys? Esp. can I still use the "normal" DF limit for the metadata queries (which basically only have a filter, no grouping, no projection).

Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:457]: It isn't obvious, but `LIMIT` and `OFFSET` are applied per _group key_ across all statements. The group key is derived from the point (row) being produced and is defined as the measurement name followed by any tags specified in a `GROUP BY` clause. Given only the `SELECT` statement can specify a `GROUP BY`, that means only the `measurement` applies for the `SHOW` statements.

It isn't obvious, but the measurement grouping applies to `SHOW TAG (KEYS|VALUES)` and `SHOW FIELD KEYS`. I speak about how InfluxQL data flow works in #6112, specifically the section titled **Grouping behaviour**[^no_link].

[^no_link]: Unfortunately, you can't link to headings in an issue :(

**Comment by stuartcarnie** [comment]
Thanks for the review, @mhilton! Fixed the clippy issue and specifically ignored the `too-many-arguments` for `project_select` for now as I expect it will be refactored when implementing sub-queries

**Comment by stuartcarnie** [comment]
Also note that I identified an issue with the output per

- https://github.com/influxdata/influxdb_iox/issues/7460

File: iox_query_influxql/src/plan/planner.rs (hunk 1 above)
Comment (crepererum) [reply iox_query_influxql/src/plan/planner.rs:458]: Thank you, that helps a lot :)

**Pull request merged**